### ISSUE TITLE
B should cancel segment processing if the HTTP push client request ends

### ISSUE THREAD

**Issue by yondonfu** [livepeer/go-livepeer#2113]
**Is your feature request related to a problem? Please describe.**

A HTTP push client may close its connection or cancel a request before B finishes processing a segment. At the moment, we [check if the request context is cancelled](https://github.com/livepeer/go-livepeer/blob/b96a0f5fc690740b8e2b8cc79d67992c7d90d8f3/server/mediaserver.go#L966), but only after B finishes processing a segment. As a result, B could continue processing a segment even though the client has already given up on the request and is no longer listening for the response.

Consider, the following example with Mist server as the HTTP push client:

- Mist sets the upload timeout for a segment to segment duration + 2s resulting in a 4s timeout for a 2s segment
- Mist sends the segment to B
- B starts processing the segment and gets stuck waiting for an upload to O
- Mist times out after 4s
- B times out waiting for an upload to O after 8s (the timeout is currently set to 4x the segment duration) and then retries 3 times meaning that it could spend 8s * 3 = 24s processing the segment if there are networking issues with each O
- B completes segment processing and is ready to return an error to Mist, but Mist already stopped listening for a response

**Describe the solution you'd like**

If the HTTP push client cancels its request, B should immediately cancel segment processing because the client is no longer listening for a response.

Here is a sketch of one way we could implement this:

```golang
func processSegmentWithContext(ctx context.Context, cxn *rtmpConnection, seg *stream.HLSSegment) ([]string, error) {
    resCh := make(chan struct {
        urls []string
        err error
    })

    go func() {
        urls, err := processSegment(cxn, seg)
        resCh <- struct{urls []string; err error}{urls, err}
    }()

    select {
        case <-ctx.Done():
            return nil, ctx.Err()
        case res := <-resCh:
            return res.urls, res.err
    }
}
```

We may want to pass a context to `transcodeSegment()` as well?

**Describe alternatives you've considered**

N/A

**Additional context**

N/A

**Comment by leszko** [comment]
I tried to reproduce this issue by pushing segments with the following command.

```
ffmpeg -re -f lavfi -i testsrc=size=1280x720:rate=30 -f lavfi -i sine -vf drawtext="text='%{localtime\:%X}':fontsize=20:fontcolor=white:x=7:y=7" -pix_fmt yuv420p -vcodec libx264 -vb 2000k -preset ultrafast -acodec aac -f hls -method POST http://localhost:8936/live/movie/
```

What I notice is that every segment times out [here](https://github.com/livepeer/go-livepeer/blob/b96a0f5fc690740b8e2b8cc79d67992c7d90d8f3/server/mediaserver.go#L967). I'm not sure about the reason, maybe ffmpeg just has a too short timeout for the HTTP connection. But in any case, I have two questions:

1. How to reproduce the description of this GH Issue (any other ffmpeg command or better to use the mist server)?

2. If we change the logic as suggested in the description, then the command I posted won't be transcoded at all. Isn't it an issue? Currently, even if HTTP upload times out or is canceled, the segment is still transcoded, so the user can read it. IIUC after this change, the segment won't be transcoded at all.

CC: @yondonfu @darkdarkdragon @jailuthra

**Comment by leszko** [comment]
Discussed offline with @yondonfu . What we agreed on is that the most important use case is the MistServer and that a client pushing HTTP segments to the broadcasters should always wait for the HTTP response. 

Therefore what we plan to do is to:
- Implement the logic as originally described in this GH Issue
- Try to find a solution for `ffmpeg` still work (maybe add an additional parameter or maybe it's possible to force `ffmpeg` to wait for the HTTP response)

**Comment by yondonfu** [comment]
Some more context:

- At the moment, a client can push a segment and synchronously wait for the multi-part response *or* it could also just query the playlist from the B which is updated as segments are transcoded
- The former is the main way that the HTTP push API is used today by Mist
- But, the latter is how you can easily test the HTTP push API using ffmpeg
- The main upside of cancelling processing after a client disconnects is that it reduces overall load on the system
- However, if we update B to stop processing a segment after the client disconnects, then that means that you a client can't just push a segment and expect to get an updated playlist separately
- One potential way to preserve that behavior is allowing the client to specify a deadline for the request so that the client can push a segment and disconnect with B processing the segment until the deadline (there should probably be some reasonable default behavior so that you can't just set an arbitrarily long deadline)

**Comment by iameli** [comment]
Seems like perhaps the `rw_timeout` option might be pertinent here? https://ffmpeg.org/ffmpeg-protocols.html#Protocols

**Comment by Thulinma** [comment]
It would make sense to me to always make a "reasonable attempt" to complete the transcode if a segment was (fully?) received, regardless of client connection status. It's a common pattern in HTTP services to be "idempotent" in the sense that requesting the same segment to be transcoded "again" does not start a new transcode, but immediately returns the result of the previous transcode attempt (if available, or waits for it if still in progress).

Ideally, there is an early response (like WebDAV's `102 Processing` response) that indicates the request was received and is being processed, so that the client knows the broadcaster is responsive and operational. The reason we have a relatively low timeout in Mist is to reduce the delay incurred by a potential malfunctioning broadcaster. If we have a way to determine the broadcaster is operational, we can reasonably increase the timeout after receiving that information. Even better if we can get regular updates on the transcode process!

This could of course also be done by using a different protocol than plain HTTP (for example WebSockets) and/or switching to a full-streaming workflow. 🤔

**Comment by iameli** [comment]
Right - I'd be nervous about introducing any assumptions about statefulness in the current workflow. In production there's no guarantee that any kind of POST is going to hit the same actual node behind the scenes. A fully-streaming workflow could make different assumptions, of course.

**Comment by Thulinma** [comment]
Idempotent does not mean stateful, though! The work has already started and is likely tricky to cancel in most cases; so it makes sense to complete it and keep it around for a while if possible/practical. If the next request hits another node, well, everything still works. And if it hits the same one; you gain quite a bit of efficiency. ¯\\\_(ツ)\_/¯

**Comment by leszko** [comment]
Thanks for the comments. I see we have more options to consider :)

@iameli what you mentioned about the statelessness, so you mean we should treat the broadcaster like a stateless service that can run behind a (round-robin) load-balancer? It sounds somehow counter-intuitive to me since that means that the stream can bounce between different broadcasters.

@Thulinma You mentioned that `If we have a way to determine the broadcaster is operational, we can reasonably increase the timeout after receiving that information.`. If the broadcaster is malfunctioning, shouldn't it all of the sudden return an error?

It all makes me think that maybe we could keep the current behavior, just improve the error propagation and fail-fast approach 🤔 If we know that the orchestrator is malfunctioning, it should all of a sudden return an error. Then, Mist could increase the timeout...

CC: @yondonfu

**Comment by victorges** [comment]
Chiming in as well...

> We may want to pass a context to processSegment() as well?

Yes, passing the context down to the function that actually does the work (e.g. spawns ffmpeg, calls some other service) is crucial, otherwise we will make a "shallow" cancellation in the end but the core work will continue to be performed the same way in background. The (or a child) `ctx` must be propagated down to the lowest level of I/O performed so the cancellation actually stops any processing.

> Ideally, there is an early response (like WebDAV's 102 Processing response) that indicates the request was received and is being processed.

Yeah! I like that too. It could be done with pure HTTP codes as well like a `202 Accepted` while the work keeps happening in background. It could also be toggled by a query string or something like that. 
e.g.: In the API handler sth like this:
```
if req.URL.Query().Get("async") == "1" {
    ctx = context.Background()
    go processSegment(ctx)
    res.WriteHeaders(http.StatusAccepted)
    return
}
err := processSegment(ctx)
// handle...
res.WriteHeaders(http.StatusOK)
```

> It would make sense to me to always make a "reasonable attempt" to complete the transcode if a segment was (fully?) received, regardless of client connection status.

I also like this since, for example, what happens in someone does a "cancellation attack" and the public network orchestrators get several requests that get cancelled in the end and they are not paid for? Maybe we should avoid cancelling those requests at all costs. 

Maybe all we need is to check if the context is cancelled before proceeding with Orchestrator transcode retries? In the condition of this loop: https://github.com/livepeer/go-livepeer/blob/02460b9c84ed6a6b46434ae7bd8a3f488c9f14c4/server/broadcast.go#L796

> If the broadcaster is malfunctioning, shouldn't it all of the sudden return an error?

I think it could happen that the logic has some kind of problem, gets stuck, deadlocks or something like that. So I think it's healthy to always have a timeout in service-to-service calls, but they could probably be looser everywhere. I'd say that timeout of a caller should **always** be greater or equal to then the cumulated timeout of a well-functioning execution of the callee. In this case, mist timeout should be at least 3x the timeout of the broadcaster with orchestrators IMO.

**Comment by yondonfu** [comment]
> Ideally, there is an early response (like WebDAV's 102 Processing response) that indicates the request was received and is being processed.

We could follow the same flow as the [B <> O protocol](https://github.com/livepeer/go-livepeer/blob/master/doc/networking.md#post-segment) in which O [returns a 200 OK header](https://github.com/livepeer/go-livepeer/blob/master/server/segment_rpc.go#L167) when it finishes downloading the segment data. An additional benefit of this is that the push client (i.e. Mist) would then have the ability to define more granular timeouts if needed i.e. a timeout for the segment upload + a timeout for the segment processing. If the client gets a timeout during the segment upload then it is likely that the problem is with the broadcaster as opposed to with an orchestrator.

> It could also be toggled by a query string or something like that

Interesting idea. So, this would be formalizing a sync vs. async way to use the push API. The sync way would be for the client to wait for the multi-part response. The async way would be for the client to push the segment and then async query for the playlist later.

The question remains of when to cancel the context that is passed to the segment processing functions to prevent further unnecessary transcode attempts. If we formalize a sync vs. async way to use the push API, then in the sync flow a disconnected client could cancel the context and in the async flow we could skip context cancellation initially and consider adding a deadline timeout for async requests separately (since this isn't really a flow used in prod right now).

> Maybe all we need is to check if the context is cancelled before proceeding with Orchestrator transcode retries?

I think this makes sense. B would let an already running transcode attempt complete, but if it sees that the context is cancelled then it will not bother with another transcode attempt.

> I'd say that timeout of a caller should always be greater or equal to then the cumulated timeout of a well-functioning execution of the callee.

👍 We currently don't have this property because for a 2s segment Mist will set the timeout to 4s while the timeout set by a B for its communication with a O will be 8s.

**Comment by Thulinma** [comment]
> @Thulinma You mentioned that `If we have a way to determine the broadcaster is operational, we can reasonably increase the timeout after receiving that information.`. If the broadcaster is malfunctioning, shouldn't it all of the sudden return an error?
> 
> It all makes me think that maybe we could keep the current behavior, just improve the error propagation and fail-fast approach thinking If we know that the orchestrator is malfunctioning, it should all of a sudden return an error. Then, Mist could increase the timeout...

Yes, but (as Victor already mentioned) it's not good practice to rely on the other side to report errors; as an endpoint that is malfunctioning is likely not able to relay that information. If it _can_ relay it, yeah, definitely do so. We just can't _depend_ on that mechanism working.

> Yeah! I like that too. It could be done with pure HTTP codes as well like a `202 Accepted` while the work keeps happening in background.

That works too, yes. 👍

> I'd say that timeout of a caller should **always** be greater or equal to then the cumulated timeout of a well-functioning execution of the callee. In this case, mist timeout should be at least 3x the timeout of the broadcaster with orchestrators IMO.

In principle, yeah, I agree with you. But in this case Mist also has a higher purpose for the transcode: delivering a working stream to end-users. And the current delay (segment duration + 2s) is already significantly longer than it should be to provide reliable service to end-users. It might actually make more sense to abort and simply fill the stream with placeholder frames instead at that point. 🤔 Ideally I'd like to be able to abort in _significantly less_ than the segment duration, so the retries actually have a chance of still getting the job done in time to prevent buffering on the viewer side.

> An additional benefit of this is that the push client (i.e. Mist) would then have the ability to define more granular timeouts if needed i.e. a timeout for the segment upload + a timeout for the segment processing. If the client gets a timeout during the segment upload then it is likely that the problem is with the broadcaster as opposed to with an orchestrator.

Mist actually already can (and does) do this, right now. 😎

**Comment by leszko** [comment]
Thanks for all the comments. I really like what @yondonfu and @victorges wrote about using the same strategy as we do in the B<>O protocol and "check if the context is canceled before proceeding with Orchestrator transcode retries". I think It solves all the issues.

Then, it would work as follows:
1. Send a segment: Mist -> O
2. When the segment is uploaded, suddenly return HTTP 200 OK
3. Start Transcoding, Mist can wait on reading the body response stream
4. If Mist times out, it closes the connection
5. When the connection is closed, we just do not retry additional attempts B<>O (but we don't actively cancel anything)

With this approach, FFMPEG can still work correctly and we don't have any issues with "statefulness", since everything is within one HTTP connection. Then, I think we don't need this async push API for now. We can always implement it in the future.

What do you think? If we agree on this solution, I'll send a PR.

**Comment by victorges** [comment]
> It might actually make more sense to abort and simply fill the stream with placeholder frames instead at that point. 🤔
 
I actually like that approach! Is there any possibility to fill a placeholder after some specified timeout, but still keep the transcode attempt running (and maybe even retries) so that if it eventually succeeds we can go back and patch the stream with the transcoded segments? Also any possibility to fallback to source-only when transcoding is not working, even for players/clients that had already started playing a transcoded rendition?

> What do you think? If we agree on this solution, I'll send a PR.

I think it is a good way to go, I would just add a nit on the point `2.`: We don't need to return an early status code/headers, since the timeout of the client for the body is very most likely the same as the timeout of the headers. But if we do respond the headers already, we will lose the ability to return error status code/headers when something goes wrong in further processing of the request, which is sub-ideal IMO. 

As some examples, any proxy on the way (e.g. traefik) might stop logging/counting those responses as errors, and we would be incompatible with tools that use the status code to measure the success rate of given services (traefik again, which sends some metrics to prometheus I'd love to start using soon). I think the early return on that case is only really necessary if we do actually have a multi-request protocol, in which after receiving a "partial success" code the client will start polling for the result of that long-running operation. It's generally a better practice to avoid a long-running connection having to be open, but since we are keeping that connection anyway I think we don't need the early return of the status code (since we don't know what it will effectively be).

**Comment by yondonfu** [comment]
> But if we do respond the headers already, we will lose the ability to return error status code/headers when something goes wrong in further processing of the request, which is sub-ideal IMO.

Fair point. For the B <> O protocol, the expectation right now is that O will return an early 200 OK once the segment is uploaded and then B should parse the response later on for any errors that might've occurred during transcoding. This does introduce the inconvenience of potential incompatibility with tools like proxies that rely exclusively on headers for error management. Since the use of such tools for a push client <> B may be more prevalent right now and since we'd need to make sure that Mist continues waiting for the response even after the initial 200 OK header is passed, I'd be fine with not using this pattern here for now (leaving the B <> O protocol as-is though since that's already how it works).

With that being said, I *do* like having the ability to be able to distinguish between the upload and processing phases of a request and being able to determine if the problem was during upload or if the problem was during processing. I'd advocate for separately considering updates that can provide this property *and* address the concerns with the current pattern used in the B <> O protocol.

**Comment by victorges** [comment]
>  For the B <> O protocol, the expectation right now is that O will return an early 200 OK once the segment is uploaded and then B should parse the response later on for any errors that might've occurred during transcoding. 

Oh, I see! I was not aware that this was the protocol between B <> O. In that case, I think it's very fair to implement a consistent protocol for all go-livepeer HTTP pushes. It would be nice to have a more HTTP "compliant" protocol, but maybe all this discussion will make less sense when we move to streaming architecture anyway since we won't be relying on individual request/responses (tho we will probably keep support to the old/current HTTP push protocol anyway?).

> With that being said, I do like having the ability to be able to distinguish between the upload and processing phases of a request and being able to determine if the problem was during upload or if the problem was during processing. 

In the case of error handling specifically, we could always just have a special error code that is returned when a problem happened when reading the request body on the server, either a specific status code or something custom in the response body itself like `{"error":"request_body_error"}`. But, in practice, I think that would only happen when some problem happened with the TCP connection, in which case no response would be received on the client side either.

> I'd advocate for separately considering updates that can provide this property and address the concerns with the current pattern used in the B <> O protocol.

Agreed. I think we have 3 different concerns here in the end which can be discussed and addressed separately in the end: 
 - Lower the impact of a (intentional or not) "cancellation attack" on the broadcaster. This can be ~~fixed (?)~~ improved by stopping O transcode retries if the ctx has been cancelled.
 - Make the mist <> B protocol similar to B <> O, replying with 200 as soon as request body upload finishes. I am not sure what benefits this would bring apart from the consistency, but I think that is enough reason to do it if you think it's worth.
 - Address any issues with the B<>O protocol (and potentially mist <> B). Doesn't seem like a big priority since we'll be moving to streaming anyway, but could be a nice to have depending on which protocols/tools we'd like to be compliant to.

**Comment by Thulinma** [comment]
> > It might actually make more sense to abort and simply fill the stream with placeholder frames instead at that point. thinking
> 
> I actually like that approach! Is there any possibility to fill a placeholder after some specified timeout, but still keep the transcode attempt running (and maybe even retries) so that if it eventually succeeds we can go back and patch the stream with the transcoded segments? Also any possibility to fallback to source-only when transcoding is not working, even for players/clients that had already started playing a transcoded rendition?

We can't retroactively insert any data after putting placeholders in; Mist's internals have a hard requirement on data being permanent once added to the track timeline. (At least until one "buffer duration" - the DVR window - has passed and the data is wiped from memory again.)
We _can_ fall back to source-only, and in fact that setup has been running on Picarto for months already. 😎

**Comment by leszko** [comment]
Ok, I think we've come to some consensus :) 

> * Lower the impact of a (intentional or not) "cancellation attack" on the broadcaster. This can be ~fixed (?)~ improved by stopping O transcode retries if the ctx has been cancelled.
> * Make the mist <> B protocol similar to B <> O, replying with 200 as soon as request body upload finishes. I am not sure what benefits this would bring apart from the consistency, but I think that is enough reason to do it if you think it's worth.

So, let's stick to these two points you mentioned. Make Mist<>B protocol the same as the B<>O protocol (`upload -> return HTTP 200 OK -> processing`) and do not retry if HTTP connection is closed/canceled.

> * Address any issues with the B<>O protocol (and potentially mist <> B). Doesn't seem like a big priority since we'll be moving to streaming anyway, but could be a nice to have depending on which protocols/tools we'd like to be compliant to.

Yes, let's move it aside for now.

**Comment by leszko** [comment]
> * Lower the impact of a (intentional or not) "cancellation attack" on the broadcaster. This can be ~fixed (?)~ improved by stopping O transcode retries if the ctx has been cancelled.

I've fired a PR on it: #2124, PTAL

> * Make the mist <> B protocol similar to B <> O, replying with 200 as soon as request body upload finishes. I am not sure what benefits this would bring apart from the consistency, but I think that is enough reason to do it if you think it's worth.

I've started the implementation and then had some more doubts. There is a subtle difference between the B<>O protocol and Mist<>B protocol. B<>O is a kind-of internal Livepeer protocol, so it's probably ok to suddenly return HTTP 200 OK and then stream errors in the response body. However Mist<>B protocol is more like a public API, so sticking to some HTTP standards may be important.

Wdyt? @yondonfu @Thulinma @victorges

**Comment by yondonfu** [comment]
My suggestion is to keep this issue (and the PR that closes it) focused on solving the problem of a B continuing with additional transcode attempts for a segment even after the HTTP client stops caring about the results. I believe the changes in https://github.com/livepeer/go-livepeer/pull/2124 should help address this. However, those changes would also mean that when using an HTTP client that pushes a segment to a B and then immediately disconnects (i.e. when using ffmpeg CLI without additional HTTP protocol config options), B will only attempt to transcode the segment once. I think this behavior is okay for now as a) it doesn't completely break the ability to use ffmpeg to push segments and then query for a playlist after (there will still be at least a single transcode attempt from B) and b) the workflow with ffmpeg is really used for testing at the moment so it isn't the highest priority to make it work really well and if it does become a priority we should take a more holistic approach to making it work well (as opposed to trying to shoehorn changes in right now).

Separately, I think we should track any additional changes that we want to consider for the HTTP push API (making it more similar to the B <> O protocol or otherwise) in a separate issue and continue the conversation there so we can figure out where the issue should go in backlog prioritization.

**Comment by leszko** [comment]
Goog idea @yondonfu , I created a GH Issue: https://github.com/livepeer/go-livepeer/issues/2125. We can move the discussion there. Also feel free to update the GH Issue name and description if needed.
